botwo/botwo.py

```diff
@@ -0,0 +1,188 @@
+                    kwargs["Bucket"] = map.get("mapped_bucket_name")
+                if kwargs.get("Prefix"):
+                    if map.get("key_prefix"):
+                        kwargs["Prefix"] = map.get("keu_prefix") + kwargs["Prefix"]
+                break
+        return getattr(paginator_to_call, "paginate")(**kwargs)
+
+
+class Botwo(object):
+    def __init__(self):
+        self.client1 = None
+        self.client2 = None
+        self.default_client = None
+        self.client1_name = None
+        self.client2_name = None
+        self.mapping = None
+        self.meta = None
+
+    def create_client_class(self, client_mapping, config):
+        if type(client_mapping) is not dict:
+            raise TypeError("Invalid client mapping type: " + str(type(client_mapping)) + " expected dict")
+        if len(client_mapping) != 2:
+            raise TypeError("two clients is expected")
+
+        names = list(client_mapping)
+        self.client1_name = names[0]
+        self.client1 = client_mapping.get(names[0])
+        self.client2_name = names[1]
+        self.client2 = client_mapping.get(names[1])
+
+        self.mapping = self._parse_config(config)
+        class_attributes = self._create_methods()
+
+        bases = [Botwo]
+        cls = type("s3", tuple(bases), class_attributes)
+        return cls
+
+    def _parse_config(self, config):
+        if len(config) == 0:
+            raise ValueError("default client is not provided")
+        if config[-1].get("bucket_name") == "*":
+            if config[-1].get("client") == self.client1_name:
+                self.default_client = self.client1
+            elif config[-1].get("client") == self.client2_name:
+                self.default_client = self.client2
+        else:
+            raise ValueError("default client is not provided")
+        self.meta = self.default_client.meta
+        return config
+
+    def _create_methods(self):
+        op_dict = {}
+        operations = [func for func in dir(self.default_client) if
+                      (callable(getattr(self.default_client, func)) and not func.startswith('_'))]
+        for operation_name in operations:
+            if operation_name in SPECIAL_METHODS:
+                op_dict[operation_name] = self._create_special_method(operation_name)
+            elif operation_name in COPY_METHODS:
+                op_dict[operation_name] = self._create_copy_method(operation_name)
+            else:
+                op_dict[operation_name] = self._create_api_method(operation_name)
+        return op_dict
+
+    def _create_api_method(self, operation_name):
+        def _api_call(*args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            client_to_call = self.default_client
+            if kwargs.get("Bucket"):  # bucket operation
+                full_path = kwargs["Bucket"] + "/" + kwargs["Key"]
```

Comment:
IMO you should never concatenate bucket and key. It's a bug waiting to happen

Reply:
Some operations operate on a `Prefix`, rather than a key (e.g. list objects operations). This will result in a KeyError.

Reply:
Done